x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/normalizer/ShortCircuitingRenormalizer.java

```diff
@@ -116,37 +116,53 @@ private synchronized AugmentedQuantiles getLatestAugmentedQuantilesAndClear() {
     }
 
     private synchronized boolean tryStartWork() {
-        if (phaser.getUnarrivedParties() > 0) {
+        if (latestQuantilesHolder == null) {
             return false;
         }
-        return phaser.register() >= 0;
+        // Don't start a thread if another normalization thread is still working. The existing thread will
+        // do this normalization when it finishes its current one. This means we serialise normalizations
+        // without hogging threads or queuing up many large quantiles documents.
+        if (semaphore.tryAcquire()) {
+            try {
+                latestTask = executorService.submit(this::doRenormalizations);
+            } catch (RejectedExecutionException e) {
+                latestQuantilesHolder.getLatch().countDown();
+                latestQuantilesHolder = null;
+                latestTask = null;
+                semaphore.release();
+                logger.warn("[{}] Normalization discarded as threadpool is shutting down", jobId);
+                return false;
+            }
+            return true;
+        }
+        return false;
     }
 
     private synchronized boolean tryFinishWork() {
         // Synchronized because we cannot tolerate new work being added in between the null check and releasing the semaphore
         if (latestQuantilesHolder != null) {
             return false;
         }
-        phaser.arriveAndDeregister();
+        semaphore.release();
+        latestTask = null;
         return true;
     }
 
     private void doRenormalizations() {
         do {
             AugmentedQuantiles latestAugmentedQuantiles = getLatestAugmentedQuantilesAndClear();
             assert latestAugmentedQuantiles != null;
```

Comment:
Just because we do an assert, production runs with asserts off. We should still prevent an NPE.

Reply:
My thinking was that the logic is now simple enough that it cannot be `null`.  In the code before #85555 there was an expected way that it could be `null`.  In the code of #85555 it really should never have been `null`, but the `Phaser` was so complex that I wasn't convinced.  Whereas now I really cannot see any way that it could be `null`.

I guess it doesn't hurt to add an extra `if`.  But I'll also add a `TODO` to remove it if the `assert` doesn't trip in the next year or so.

Reply:
I see that as well. THe logic now is much simpler and easier to understand. ++ refactor from me!